x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,430 @@
+
+	// The duration (in seconds) for which the call waits for a message to arrive
+	// in the queue before returning. If a message is available, the call returns
+	// sooner than WaitTimeSeconds. If no messages are available and the wait time
+	// expires, the call returns successfully with an empty list of messages.
+	waitTimeSecond int64 = 10
+)
+
+func init() {
+	err := input.Register(inputName, NewInput)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Input is a input for s3
+type Input struct {
+	outlet     channel.Outleter // Output of received s3 logs.
+	config     config
+	awsConfig  awssdk.Config
+	logger     *logp.Logger
+	close      chan struct{}
+	workerOnce sync.Once // Guarantees that the worker goroutine is only started once.
+}
+
+type s3Info struct {
+	name   string
+	key    string
+	region string
+	arn    string
+}
+
+type sqsMessage struct {
+	Records []struct {
+		EventSource string `json:"eventSource"`
+		AwsRegion   string `json:"awsRegion"`
+		EventName   string `json:"eventName"`
+		S3          struct {
+			Bucket struct {
+				Name string `json:"name"`
+				Arn  string `json:"arn"`
+			} `json:"bucket"`
+			Object struct {
+				Key string `json:"key"`
+			} `json:"object"`
+		} `json:"s3"`
+	} `json:"Records"`
+}
+
+// NewInput creates a new s3 input
+func NewInput(cfg *common.Config, outletFactory channel.Connector, context input.Context) (input.Input, error) {
+	cfgwarn.Beta("s3 input type is used")
+
+	logger := logp.NewLogger(inputName)
+
+	config := defaultConfig
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "failed unpacking config")
+	}
+
+	outlet, err := outletFactory(cfg, context.DynamicFields)
+	if err != nil {
+		return nil, err
+	}
+
+	awsConfig, err := awscommon.GetAWSCredentials(config.AwsConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "getAWSCredentials failed")
+	}
+
+	p := &Input{
+		outlet:    outlet,
+		config:    config,
+		awsConfig: awsConfig,
+		logger:    logger,
+		close:     make(chan struct{}),
+	}
+
+	return p, nil
+}
+
+func (c *config) Validate() error {
+	if c.VisibilityTimeout < 0 || c.VisibilityTimeout.Hours() > 12 {
+		return errors.New("visibilityTimeout is not defined within the " +
+			"expected bounds")
+	}
+	return nil
+}
+
+// Run runs the input
+func (p *Input) Run() {
+	p.workerOnce.Do(func() {
+		p.logger.Infof("s3 input worker has started. with queueURL: %v", p.config.QueueURL)
+
+		visibilityTimeout := int64(p.config.VisibilityTimeout.Seconds())
+		regionName, err := getRegionFromQueueURL(p.config.QueueURL)
+		if err != nil {
+			p.logger.Errorf("failed to get region name from queueURL: %v", p.config.QueueURL)
+		}
+
+		awsConfig := p.awsConfig.Copy()
+		awsConfig.Region = regionName
+		svcSQS := sqs.New(awsConfig)
+		svcS3 := s3.New(awsConfig)
+
+		go p.run(svcSQS, svcS3, visibilityTimeout)
+		go p.runKeepAlive()
+	})
+}
+
+func (p *Input) run(svcSQS *sqs.Client, svcS3 *s3.Client, visibilityTimeout int64) {
+	for {
```

Comment:
We need to ensure this loop finishes, I think you want the content of `runKeepAlive` here instead of in another function?

Reply:
```
for {
		select {
		case <-p.close:
			return
		default:
		}
                  // receive messages from sqs
		req := ...
                 ...
                 p.processor(...)
}
```
I tried this but that won't let me exit when I tried to kill filebeat hmmm 
Is it because I'm checking `p.close` which is a channel, so I need to run `p.runKeepAlive` as separate goroutine in order to check `p.close`?